server/documents/document.js

```diff
@@ -44,6 +48,291 @@ Meteor.methods({
+        // eslint-disable-next-line no-console
+        console.error("Error applying a step.", result.failed);
+        throw new Meteor.Error('invalid-request', "Invalid step.");
+      }
+      doc = result.doc;
+      version = content.version;
+    });
+
+    // Get original document steps that were applied after fork.
+    const originalSteps = Content.documents.find(
+      {
+        version: {
+          $gt: fork.forkedAtVersion,
+        },
+        contentKeys:
+        {
+          $elemMatch: {
+            $in: [original.contentKey],
+          },
+        },
+      },
+      {
+        sort: {
+          version: 1,
+        },
+      },
+    ).fetch()
+    .map((x) => {
+      return Step.fromJSON(schema, x.step);
+    });
+
+    const shouldRebase = transform !== undefined && forkSteps.length > 0;
+
+    if (shouldRebase) {
+      // Revert steps that were applied on the original document after fork.
+      for (let i = transform.steps.length - 1; i >= 0; i -= 1) {
+        const result = transform.steps[i].invert(transform.docs[i]).apply(doc);
+        transform.step(transform.steps[i].invert(transform.docs[i]));
+        if (!result.doc) {
+          // eslint-disable-next-line no-console
+          console.error("Error applying a step.", result.failed);
+          throw new Meteor.Error('invalid-request', "Invalid step.");
+        }
+        doc = result.doc;
+      }
+    }
+    else {
+      transform = new Transform(doc);
+    }
+
+    // Apply all the forked document steps.
+    for (let i = 0; i < forkSteps.length; i += 1) {
+      const result = forkSteps[i].apply(doc);
+      transform.step(forkSteps[i]);
+      if (!result.doc) {
+        // eslint-disable-next-line no-console
+        console.error("Error applying a step.", result.failed);
+        throw new Meteor.Error('invalid-request', "Invalid step.");
+      }
+      doc = result.doc;
+    }
+
+    if (shouldRebase) {
+      // Remap original document steps and apply.
+      for (let i = 0, mapFrom = originalSteps.length * 2; i < originalSteps.length; i += 1) {
+        const mapped = originalSteps[i].map(transform.mapping.slice(mapFrom));
+        mapFrom -= 1;
+        if (mapped && !transform.maybeStep(mapped).failed) {
+          const result = mapped.apply(doc);
+          transform.mapping.setMirror(mapFrom, transform.steps.length - 1);
+
+          if (!result.doc) {
+            // eslint-disable-next-line no-console
+            console.error("Error applying a step.", result.failed);
+            throw new Meteor.Error('invalid-request', "Invalid step.");
+          }
+          doc = result.doc;
+        }
+      }
+    }
+
+    const timestamp = new Date();
+
+    // Save merge steps.
+    transform.steps.forEach((x, i) => {
+      if (i >= originalSteps.length) {
+        version += 1;
+        Content.documents.upsert({
+          version,
+          contentKeys: {
+            $elemMatch: {
+              $in: [original.contentKey],
```

Comment:
Hmm. Are you sure this is correct?

Where do you remove steps done on the original document after the fork point? Or do you add both revert steps and then rebased steps?

Why are you using here upsert if you are not really updating anything?

Reply:
Yes I do, I add both revert steps and then rebased steps. I'll use insert instead of upsert.